apps/src/dance/lottie/LottieDancerRenderer.ts

```diff
@@ -177,60 +186,194 @@ export default class LottieDancerRenderer {
+    this.skeletonNamePromise = (async () => {
+      if (!skeletonParam && this.bodyMetadataUrl) {
+        try {
+          const bodyMetaData = await fetchJson<{skeleton?: string}>(
+            this.bodyMetadataUrl
+          );
+          const skeletonNameFromJson = bodyMetaData?.skeleton;
+          if (typeof skeletonNameFromJson === 'string') {
+            console.log(
+              `Resolved skeleton name "${skeletonNameFromJson}" from body metadata JSON.`
+            );
+            return skeletonNameFromJson.trim().toLowerCase();
+          }
+        } catch {}
+      }
+      return skeletonParam || TEST_SKELETON;
+    })();
+    return this.skeletonNamePromise;
+  }
+
   /**
    * Load, recolor, inject head, and memoize a move's Lottie JSON.
    * Returns the cached/transformed JSON if already present.
    */
   private async loadAndTransformMove(danceMove: string): Promise<LottieJSON> {
+    const skeletonName = await this.getSkeletonName();
+    const key = danceMove.toLowerCase();
     const danceMoveLowerCase = String(danceMove).toLowerCase();
     if (this.cachedAnimationData[danceMoveLowerCase]) {
       return this.cachedAnimationData[danceMoveLowerCase];
     }
-
-    const jsonUrl = resolveAnimationUrl(this.skeletonName, danceMoveLowerCase);
-
-    const animData = await fetchJson<LottieJSON>(jsonUrl);
-
-    // Fetch palette metadata if we have a URL for it.
-    let palette: Palette | null = null;
-    if (this.metadataUrl) {
-      const metadataJson = await fetchJson<DancerMetadata>(this.metadataUrl);
-      palette = normalizePalette(metadataJson);
+    const pendingLoad = this.pendingAnimationLoads.get(key);
+    if (pendingLoad) {
+      return pendingLoad;
     }
-
-    // Recolor assets based on hard-coded accessory-name rules.
-    applyColorMapping(animData, palette);
-
-    // Replace vector head with an image, when head.png is available.
-    const headInfo = await fetchHeadImageInfo(this.headUrl);
-    if (headInfo) {
-      const headPre = findHeadPrecompLayerDeep(animData);
-      if (headPre?.refId) {
-        const headComp = getAssetById(animData, headPre.refId);
-        if (headComp && Array.isArray(headComp.layers)) {
-          const {insertIndex, headKs} = hideVectorHeadInComp(headComp);
-          const assetId = ensureHeadImageAsset(
-            animData,
-            headInfo.dataUrl,
-            headInfo.width,
-            headInfo.height
-          );
-          insertHeadImageLayer(
-            headComp,
-            insertIndex,
-            assetId,
-            headInfo.width,
-            headInfo.height,
-            this.headScale,
-            headKs
+    const loadPromise = (async () => {
+      const jsonUrl = resolveAnimationUrl(skeletonName, danceMoveLowerCase);
+
+      const animData = await fetchJson<LottieJSON>(jsonUrl);
+      const shorten = (url?: string) =>
+        url?.match(/generate\/([^?]+)/)?.[1] || url;
+
+      console.log('Creating Lottie Dancer with:', {
```

Comment:
This is intentional. I think as we continue to test and try out different heads/bodies/metadata, we'll want a convenient way to quickly find assets that could be problematic. This could also help teachers to help us to identify anything problematic.